llama-index-integrations/llms/llama-index-llms-nvidia/llama_index/llms/nvidia/base.py

```diff
@@ -0,0 +1,379 @@
+            api_key=api_key,
+            base_url=BASE_CATALOG_URL,
+            timeout=timeout,
+            max_retries=max_retries,
+        )
+        self._aclient._custom_headers = {"User-Agent": "llama-index-llms-nvidia"}
+
+        super().__init__(
+            model=model,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            timeout=timeout,
+            max_retries=max_retries,
+            callback_manager=callback_manager,
+            system_prompt=system_prompt,
+            messages_to_prompt=messages_to_prompt,
+            complettion_to_prompt=complettion_to_prompt,
+            pydantic_program_mode=pydantic_program_mode,
+            output_parser=output_parser,
+        )
+
+    @property
+    def available_models(self) -> List[Model]:
+        ids = API_CATALOG_MODELS.keys()
+        if self._mode == "nim":
+            ids = [model.id for model in self._client.models.list()]
+        return [Model(id=name) for name in ids]
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "NVIDIA"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        params = {
+            "num_output": self.max_tokens,
+            "is_chat_model": True,
+            "model_name": self.model,
+        }
+        if context_window := catalog_modelname_to_contextsize(self.model):
+            params["context_window"] = context_window
+        return LLMMetadata(**params)
+
+    @property
+    def _model_kwargs(self) -> Dict[str, Any]:
+        return {
+            "model": self.model,
+            "temperature": self.temperature,
+            "max_tokens": self.max_tokens,
+        }
+
+    def mode(
+        self,
+        mode: Optional[Literal["nvidia", "nim"]] = "nvidia",
+        *,
+        base_url: Optional[str] = None,
+        model: Optional[str] = None,
+        api_key: Optional[str] = None,
+    ) -> "NVIDIA":
+        """
+        Change the mode.
+
+        There are two modes, "nvidia" and "nim". The "nvidia" mode is the default
+        mode and is used to interact with hosted NIMs. The "nim" mode is used to
+        interact with NVIDIA NIM endpoints, which are typically hosted on-premises.
+
+        For the "nvidia" mode, the "api_key" parameter is available to specify
+        your API key. If not specified, the NVIDIA_API_KEY environment variable
+        will be used.
+
+        For the "nim" mode, the "base_url" parameter is required and the "model"
+        parameter may be necessary. Set base_url to the url of your local NIM
+        endpoint. For instance, "https://localhost:9999/v1". Additionally, the
+        "model" parameter must be set to the name of the model inside the NIM.
+        """
+        # if mode == "nvidia":
+        #     if not api_key:
+        #         warnings.warn(
+        #             "'nvidia' mode without an api_key may result in an error",
+        #             UserWarning,
+        #         )
```

Comment:
Any reason this is commented out? If its not needed, we can remove the comments

Reply:
not a good enough one to keep it, i'll remove this